replica/table.cc

```diff
@@ -3515,11 +3530,20 @@ future<> table::clear_inactive_reads_for_tablet(database& db, storage_group& sg)
     }
 }
 
-future<> table::stop_compaction_groups(storage_group& sg) {
+future<> storage_group::stop() {
+    if (_async_gate.is_closed()) {
+        co_return;
+    }
     // Synchronizes with in-flight writes if any, and also takes care of flushing if needed.
-    for (auto& cg_ptr : sg.compaction_groups()) {
+    for (auto& cg_ptr : compaction_groups()) {
```

Comment:
A compaction group can be stopped while `table::parallel_foreach_compaction_group` still uses it

Reply:
This is similar to shutdown and already handled with stop(), which closes gates, stops ongoing compaction and waits for flush if any. I also made compaction_group a shared_ptr to avoid lifetime issues.